server/src/main/scala/io/delta/standalone/internal/DeltaSharedTableLoader.scala

```diff
@@ -530,89 +537,169 @@ class DeltaSharedTable(
+      .getOrElse(latestVersion)
+      .min(latestVersion)
     val timestampsByVersion = DeltaSharingHistoryManager.getTimestampsByVersion(
       deltaLog.store,
       deltaLog.logPath,
-      startingVersion,
-      latestVersion + 1,
+      start,
+      end + 1,
       conf
     )
 
+    // Enforce page size only when `maxFiles` is specified for backwards compatibility.
+    val pageSizeOpt = maxFilesOpt.map(_.min(queryTablePageSizeLimit))
+    val tokenGenerator = { (v: Long, idx: Int) =>
+      val nextPageTokenStr = encodeQueryTablePageToken(
+        QueryTablePageToken(
+          id = Some(tableConfig.id),
+          startingVersion = Some(v),
+          endingVersion = Some(end),
+          latestVersion = Some(latestVersion),
+          checksum = Some(queryParamChecksum),
+          startingActionIndex = Some(idx),
+          expirationTimestamp = Some(System.currentTimeMillis() + queryTablePageTokenTtlMs)
+        )
+      )
+      model.NextPageToken(nextPageTokenStr).wrap
+    }
+    var numSignedFiles = 0
     val actions = ListBuffer[Object]()
-    deltaLog.getChanges(startingVersion, true).asScala.toSeq
-      .filter(_.getVersion <= latestVersion).foreach{ versionLog =>
-      val v = versionLog.getVersion
-      val versionActions = versionLog.getActions.asScala.map(x => ConversionUtils.convertActionJ(x))
-      val ts = timestampsByVersion.get(v).orNull
-      versionActions.foreach {
-        case a: AddFile if a.dataChange =>
-          actions.append(
-            getResponseAddFile(
-              a,
-              fileSigner.sign(absolutePath(deltaLog.dataPath, a.path)),
-              v,
-              ts.getTime,
-              responseFormat,
-              true
-            )
-          )
-        case r: RemoveFile if r.dataChange =>
-          actions.append(
-            getResponseRemoveFile(
-              r,
-              fileSigner.sign(absolutePath(deltaLog.dataPath, r.path)),
-              v,
-              ts.getTime,
-              responseFormat
+    deltaLog
+      .getChanges(start, true)
+      .asScala
+      .toSeq
+      .filter(_.getVersion <= end)
+      .foreach { versionLog =>
+        val v = versionLog.getVersion
+        var indexedVersionActions =
+          versionLog.getActions.asScala.map(x => ConversionUtils.convertActionJ(x)).zipWithIndex
+        val ts = timestampsByVersion.get(v).orNull
+        if (pageTokenOpt.exists(_.getStartingVersion == v)) {
+          // Skip actions that are already processed in previous pages
+          indexedVersionActions =
+            indexedVersionActions.drop(pageTokenOpt.get.getStartingActionIndex)
+        }
+        indexedVersionActions.foreach {
+          case (a: AddFile, idx) if a.dataChange =>
+            // Return early if we already have enough files in the current page
+            if (pageSizeOpt.contains(numSignedFiles)) {
+              actions.append(tokenGenerator(v, idx))
+              return actions.toSeq
```

Comment:
Question: It seems first we will generate all the changes, and then apply pagination? So pagination will not save any processing on the server side right? Is the same true for DS server in UC as well? @linzhou-db

Reply:
We're saving some process: the `start` will be equal to the `startingVersion` in the page token, so we won't unpack delta logs for versions that are already processed in previous pages.

However, we still generate all changes until the `endingVersion` each time. I guess this part can be optimized in UC. Two options I can think of: 
1. we could push down `actionListener` to `getChanges()`, so we read logs and process actions together and it'll return in the middle
2. we don't change the interface of `getChanges()`, we call it to unpack one version at a time and process version by version.

WDYT?

Reply:
I think both options should work; the first one is similar to what we do for normal query, so may be preferable for consistency reason.
In the future, we will likely process json files in parallel as well (similar to checkpoint parquet).

Let's open a ticket for now to address this in UC.

Reply:
let's have a separate discussion about how to handle this in UC.